include/irr/asset/IAsset.h

```diff
@@ -167,25 +173,81 @@ class IAsset : virtual public core::IReferenceCounted
 
         virtual core::smart_refctd_ptr<IAsset> clone(uint32_t _depth = ~0u) const = 0;
 
+		bool restoreFromDummy(IAsset* _other, uint32_t _levelsBelow)
+		{
+			assert(getAssetType() == _other->getAssetType());
+
+			if (!canBeRestoredFrom_recurseDAG(_other))
+				return false;
+
+			restoreFromDummy_impl(_other, _levelsBelow);
+			return true;
+		}
+
+		bool canBeRestoredFrom(const IAsset* _other) const
+		{
+			assert(getAssetType() == _other->getAssetType());
+
+			if (getMutability() != EM_MUTABLE)
+				return false;
+			if (_other->getMutability() != EM_MUTABLE)
+				return false;
+			if (!isADummyObjectForCache())
+				return false;
+			if (_other->isADummyObjectForCache())
+				return false;
+
+			return true;
+		}
+
+
+		inline E_MUTABILITY getMutability() const { return m_mutability; }
+		inline bool isMutable() const { return getMutability() == EM_MUTABLE; }
+		inline bool canBeConvertedToDummy() const { return !isADummyObjectForCache() && (getMutability()&EM_CPU_PERSISTENT) != EM_CPU_PERSISTENT; }
+
     protected:
-        void clone_common(IAsset* _clone) const
+		inline static bool canBeRestoredFrom_recurseDAG_call(const IAsset* _this_child, const IAsset* _other_child)
+		{
+			return _this_child->canBeRestoredFrom_recurseDAG(_other_child);
+		}
+		inline static void restoreFromDummy_impl_call(IAsset* _this_child, IAsset* _other_child, uint32_t _levelsBelow)
+		{
+			_this_child->restoreFromDummy_impl(_other_child, _levelsBelow);
+		}
+
+		virtual bool canBeRestoredFrom_recurseDAG(const IAsset* _other) const = 0;
```

Comment:
should take a `uint32_t _levelsBelow` parameter

Reply:
oh, we already discussed it

Reply:
just remind me of the corner case that would break a `_levelsBelow` parameter version?

Reply:
anything, you want to restore pipeline to the level of pipeline layout so you check only those levels but it might turn out that pipelines are incompatible on some lower level (ds layouts for example)

Reply:
ok, thx